network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -614,12 +615,174 @@ def exec(self):
 
         return QueryAnswer(True, f'{self.name1} and {self.name2} are semantically equivalent.')
 
+class SemanticDiffQuery(TwoNetworkConfigsQuery):
+    """
+    Produces a report of newly allowed and newly denied connections
+    """
+    class SingleDiff:
+        """
+        Representing a single diff between a pair of eps
+        """
+        def __init__(self, from_ep, to_ep, removed, added):
+            self.from_ep = from_ep
+            self.to_ep = to_ep
+            self.removed = removed
+            self.added = added
+
+        def __str__(self):
+            res = "\n" + str(self.from_ep) + " -> " + str(self.to_ep) + ":\n"
+            res += "\tRemoved: " + str(self.removed) + "\n"
+            res += "\tAdded: " + str(self.added)
+            return res
+
+    @staticmethod
+    def pretty_print_diff(diff_list):
+        """
+        pretty printing the results of semantic diff
+        :param diff_list:
+        :return:
+        """
+        result = ""
+        for single in diff_list:
+            if diff_list[single]:
+                result += "- " + single + str(diff_list[single]) + "\n"
+                result = result.replace('{', ' ').replace('}', ' ').replace('\'', '')
+        return result
+
+    def compute_diff(self):
+        all_diff = {}
+        # peers_to_compare |= self.disjoint_ip_blocks()
+        old_peers = self.config1.peer_container.get_all_peers_group()
+        new_peers = self.config2.peer_container.get_all_peers_group()
+        intersected_peers = old_peers & new_peers
+        removed_peers = old_peers - intersected_peers
+        added_peers = new_peers - intersected_peers
+
+        key = 'Lost connections between removed peers'
+        all_diff[key] = []
+        for pair in itertools.permutations(removed_peers, 2):
+            _, lost_conns, _ = self.config1.allowed_connections(pair[0], pair[1])
+            all_diff[key].append(
+                SemanticDiffQuery.SingleDiff(pair[0], pair[1], lost_conns, None))
+
+        key = 'Lost connections between removed peers and intersected peers'
```

Comment:
I don't think the term "intersected peers" will be meaningful to users

Reply:
changed to `persistent peers`